platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/hashmap/HalfDiskHashMap.java

```diff
@@ -422,74 +423,64 @@ public void endWriting() throws IOException {
                 storeName,
-                oneTransactionsData.size(),
+                size,
                 oneTransactionsData.stream().mapToLong(BucketMutation::size).sum());
-        // iterate over transaction cache and save it all to file
-        if (!oneTransactionsData.isEmpty()) {
-            // for each changed bucket, write the new buckets to file but do not update index yet
-            final MutableList<IntObjectPair<BucketMutation<K>>> bucketUpdates =
-                    oneTransactionsData.keyValuesView().toList();
-            final int size = bucketUpdates.size();
+
+        if (size > 0) {
             final Queue<ReadBucketResult<K>> queue = new ConcurrentLinkedQueue<>();
-            final AtomicBoolean endWritingRunning = new AtomicBoolean(true);
-            for (final IntObjectPair<BucketMutation<K>> keyValue : bucketUpdates) {
-                final int bucketIndex = keyValue.getOne();
-                final BucketMutation<K> bucketMap = keyValue.getTwo();
-                flushExecutor.execute(() -> {
-                    if (endWritingRunning.get()) {
-                        readUpdateQueueBucket(bucketIndex, bucketMap, queue);
-                    }
-                });
-            }
-            try {
-                //  write to files
-                fileCollection.startWriting();
-                int processed = 0;
-                while (processed < size) {
-                    final ReadBucketResult<K> res = queue.poll();
-                    if (res == null) {
-                        Thread.onSpinWait();
-                        continue;
-                    }
-                    if (res.error != null) {
-                        throw new RuntimeException(res.error);
-                    }
-                    try (final Bucket<K> bucket = res.bucket) {
-                        final int bucketIndex = bucket.getBucketIndex();
-                        if (bucket.getBucketEntryCount() == 0) {
-                            // bucket is missing or empty, remove it from the index
-                            bucketIndexToBucketLocation.remove(bucketIndex);
-                        } else {
-                            // save bucket
-                            final long bucketLocation = fileCollection.storeDataItem(bucket);
-                            // update bucketIndexToBucketLocation
-                            bucketIndexToBucketLocation.put(bucketIndex, bucketLocation);
-                        }
-                    } finally {
-                        processed++;
-                    }
+            final Iterator<IntObjectPair<BucketMutation<K>>> iterator =
+                    oneTransactionsData.keyValuesView().iterator();
+
+            // read and update all buckets in parallel, write sequentially in random order
+            fileCollection.startWriting();
+            int processed = 0;
+            int inFlight = 0;
```

Comment:
Yes, it would be functionally equivalent but then I'd have checking the iterator, getting task parameters and submitting it duplicated in two places. I thought having one chunk for read task submission looked cleaner.

Reply:
Code duplication can be minimized to a single method call:
```java
    submitReadUpdateBucketJob(iterator.next(), queue)
```
It implies a new private method, though. Or maybe it can be incorporated into the existing `readUpdateQueueBucket()` method.

Nevermind. I'm fine with either approach.